src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs

```diff
@@ -38,77 +38,115 @@ void IListener.Cancel()
             {
-                Events = messages.ToArray(),
-                Context = context
-            };
+                this._parent = parent;
+            }
 
-            // Single dispatch 
-            if (_singleDispatch)
+            public async Task CloseAsync(PartitionContext context, CloseReason reason)
             {
-                int len = value.Events.Length;
+                this._cts.Cancel(); // Signal interuption to ProcessEventsAsync()
 
-                Task[] dispatches = new Task[len];
-                for (int i = 0; i < len; i++)
+                // Finish listener
+                if (reason == CloseReason.Shutdown)
                 {
-                    TriggeredFunctionData input = new TriggeredFunctionData
-                    {
-                        ParentId = null,
-                        TriggerValue = value.GetSingleEventTriggerInput(i)
-                    };
-                    dispatches[i] = _executor.TryExecuteAsync(input, CancellationToken.None);
+                    await context.CheckpointAsync();
                 }
-
-                // Drain the whole batch before taking more work
-                await Task.WhenAll(dispatches);
             }
-            else
+
+            public Task OpenAsync(PartitionContext context)
             {
-                // Batch dispatch
+                return Task.FromResult(0);
+            }
 
-                TriggeredFunctionData input = new TriggeredFunctionData
+            public async Task ProcessEventsAsync(PartitionContext context, IEnumerable<EventData> messages)
+            {
+                EventHubTriggerInput value = new EventHubTriggerInput
                 {
-                    ParentId = null,
-                    TriggerValue = value
+                    Events = messages.ToArray(),
+                    Context = context
                 };
 
-                FunctionResult result = await _executor.TryExecuteAsync(input, CancellationToken.None);
-            }
-                        
-            await context.CheckpointAsync();        
-        }
+                // Single dispatch 
+                if (_parent._singleDispatch)
+                {
+                    int len = value.Events.Length;
 
-        async Task IEventProcessor.CloseAsync(PartitionContext context, CloseReason reason)
-        {
-            // Finish listener
-            if (reason == CloseReason.Shutdown)
-            {
+                    Task[] dispatches = new Task[len];
+                    for (int i = 0; i < len; i++)
+                    {
+                        Task task;
+                        if (_cts.IsCancellationRequested)
+                        {
+                            // If we stopped the listener, then we may lose the lease and be unable to checkpoint. 
+                            // So skip running the rest of the batch. The new listener will pick it up. 
+                            task = _done;
```

Comment:
What is the benefit of adding a bunch of completed Tasks? Why can't you just exit the loop if the token is cancelled (using a collection instead of an array then)?